### PR TITLE
chore: add lora metrics file support

### PR DESCRIPTION
**Pull request by Fiona-Waters**
**What this PR does / why we need it**:

Following on from training hub [PR #40](https://github.com/Red-Hat-AI-Innovation-Team/training_hub/pull/40) this PR adds LoRA metrics file support.

Summary
- Add LoRA metrics file support (training_metrics.jsonl) to the algorithm registry, enabling progression tracking for LoRA training jobs 
- Add _read_lora_sft_metrics and _transform_lora_sft methods for reading and transforming LoRA metrics to the controller-compatible progress format
- Update docstrings and comments that referenced LoRA as having no metrics

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:

Fixes # [RHOAIENG-49760](https://redhat.atlassian.net/browse/RHOAIENG-49760) 

**Verification steps** _(required; provide valid, reproducible steps for reviewers to verify your changes, e.g. setup, commands, expected outcome, notebooks)_:

Verified on team cluster via notebook using this branch. 
Screenshot 2026-03-26 at 16 12 48
Tested: 
 - addition of file when none exists 
 - removal of existing file and addition of new one
 
Can be tested using this notebook:
[test_lora_metrics.ipynb](https://github.com/user-attachments/files/26281458/test_lora_metrics.ipynb)

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

## Summary by CodeRabbit

* **New Features**
  * Added metrics tracking for LoRA SFT so controllers can read training metrics and report progress.

* **Improvements**
  * Progress reporting now rounds and clamps percentages (remains <100% until final step) and handles missing/malformed metrics gracefully.
  * Wrapper startup now removes stale single-file metrics before processing.

* **Tests**
  * Expanded tests for LoRA: metrics reading, transformation, progress clamping, startup cleanup and completeness checks.

### CODE REVIEW

File: kubeflow/trainer/rhai/traininghub.py [hunk 1]
```diff
@@ -481,6 +522,39 @@ def _transform_sft(self, metrics):
...
+            # Calculate progress based on steps
+            percent = min(100, step / max_steps * 100) if max_steps > 0 else 0
+
+            loss_val = metrics.get("loss")
+            lr_val = metrics.get("learning_rate")
+            grad_norm_val = metrics.get("grad_norm")
+
+            return {
+                "progressPercentage": int(round(percent)),
+                "estimatedRemainingSeconds": None,
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: @kapil27 This is the same for osft and sft. Is it intentional? If not I can update it in this PR. Please let me know. Thanks

File: kubeflow/trainer/rhai/traininghub.py [hunk 2]
```diff
@@ -332,6 +334,44 @@ def _read_sft_metrics(self):
...
+                # Read last line of metrics using tail
+                try:
+                    result = subprocess.run(
+                        ["tail", "-n", "1", metrics_file],
+                        capture_output=True,
+                        text=True,
+                        check=True,
+                    )
+                    last_line = result.stdout.strip()
+                except subprocess.CalledProcessError:
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:359]: This is low-risk in practice — tail -n 1 on a regular file completes instantly and won't hang

File: kubeflow/trainer/rhai/traininghub.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapil27) [reply kubeflow/trainer/rhai/traininghub.py]: No, it's not intentional, It would be great if you could apply the same fix to for those algorithms in this PR too. Thanks for offering to update it!

[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Updated to ensure progressPercentage never reaches 100 until the step actually equals max_steps/step_total.

File: kubeflow/trainer/rhai/traininghub.py [hunk 3]
```diff
@@ -332,6 +334,44 @@ def _read_sft_metrics(self):
...
+                    return {}
+
+                if last_line:
+                    return json.loads(last_line)
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:363]: The _transform_lora_sft docstring (line 534-536) explicitly documents that max_steps is included in each line of training_metrics.jsonl:

  {"step": 1, "epoch": 0.015625, "loss": 4.2727, "learning_rate": 2e-6, "max_steps": 100}

  The LoRA KubeflowProgressCallback writes max_steps into every metrics line, so _read_lora_sft_metrics already returns max_steps as part
  of the parsed JSON. The transform then reads it with metrics.get("max_steps", 0) at line 540 — no companion config file is needed. The
  existing test at line 1061 confirms this flow works correctly with max_steps in the metrics dict.

  No changes needed here.

[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:363]: Opened a PR to fix https://github.com/Red-Hat-AI-Innovation-Team/training_hub/pull/55

[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:363]: The current fallback (max_steps defaults to 0, progress reports 0) is acceptable for the short window before the upstream change lands.

File: kubeflow/trainer/rhai/traininghub.py [hunk 4]
```diff
@@ -481,6 +529,42 @@ def _transform_sft(self, metrics):
...
+        def _transform_lora_sft(self, metrics):
+            """Transform LoRA SFT schema to controller-compatible format.
+
+            LoRA metrics format (from training_metrics.jsonl):
+                {"step": 1, "epoch": 0.015625, "loss": 4.2727, "learning_rate": 2e-6,
+                 "max_steps": 100}
+            """
+            step = metrics.get("step", 0)
+            epoch = metrics.get("epoch", 0)
+            max_steps = metrics.get("max_steps", 0)
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/traininghub.py]: should we not include these metrics in the payload if they're not available, rather than setting them to 0?

**Review by robert-bell — COMMENTED** [review]
lgtm. No objection to this being included in the next release. Do we have an e2e test for this?

File: kubeflow/trainer/rhai/traininghub.py (hunk 4 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Thanks. I've updated to return None instead of 0 in the case where max_steps are not available. I have opened a [PR ](https://github.com/Red-Hat-AI-Innovation-Team/training_hub/pull/55)in the training hub repo to ensure max_steps is included in the future. 
This metrics functionality is not currently covered in any of our E2E tests. It could possibly be added to the existing LoRA test. I can create a Jira for this.

File: kubeflow/trainer/rhai/traininghub.py [hunk 5]
```diff
@@ -369,12 +411,15 @@ def _transform_osft(self, metrics):
             step_total = steps_per_epoch * total_epochs if steps_per_epoch > 0 else 0
 
             current_step_absolute = step
-            percent = min(100, (current_step_absolute / step_total * 100)) if step_total > 0 else 0
+            progress = (current_step_absolute / step_total * 100) if step_total > 0 else 0
+            percent_int = int(round(progress))
+            if current_step_absolute < step_total:
+                percent_int = min(99, percent_int)
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:418]: Updated to be capped at 100 to prevent overshoot

File: kubeflow/trainer/rhai/traininghub.py (hunk 4 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: https://redhat.atlassian.net/browse/RHOAIENG-56144

Comment (robert-bell) [reply kubeflow/trainer/rhai/traininghub.py]: Thanks @Fiona-Waters !

File: kubeflow/trainer/rhai/traininghub.py [hunk 6]
```diff
@@ -332,6 +335,44 @@ def _read_sft_metrics(self):
...
+                        check=True,
+                    )
+                    last_line = result.stdout.strip()
+                except subprocess.CalledProcessError:
+                    return {}
+
+                if last_line:
+                    return json.loads(last_line)
+
+            except FileNotFoundError:
```
Comment (briangallagher) [inline kubeflow/trainer/rhai/traininghub.py]: nit: is this needed with if not os.path.exists above?

File: kubeflow/trainer/rhai/traininghub_test.py [hunk 7]
```diff
@@ -979,6 +981,236 @@ def sample_func(lr, batch_size):
...
+
+    stale_metrics_file = ckpt_dir / "training_metrics.jsonl"
+    stale_metrics_file.write_text('{"step": 50, "loss": 1.2}\n')
+
+    # Call instrumentation directly (no exec) with port 0 for random available port
+    from kubeflow.trainer.algorithms import get_algorithm_pod_metadata
+    from kubeflow.trainer.rhai.traininghub import _create_training_hub_progression_instrumentation
+
+    algorithm_metadata = get_algorithm_pod_metadata("lora_sft")
+    apply_fn, _handler = _create_training_hub_progression_instrumentation(
```
Comment (briangallagher) [inline kubeflow/trainer/rhai/traininghub_test.py]: nit: This block to create the handler is repeated, maybe extract to common func or @pytest.fixture

File: kubeflow/trainer/rhai/traininghub.py (hunk 6 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Not needed, removed. Thanks

File: kubeflow/trainer/rhai/traininghub_test.py (hunk 7 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub_test.py]: Updated to use a @pytest.fixture. Thanks

**Review by briangallagher — APPROVED** [review]
